exercises/practice/sum-of-multiples/.approaches/from-factors/content.md

```diff
@@ -0,0 +1,103 @@
+
+In order to compute the list of multiples for a factor, we create a [`RangeFrom`][rangefrom]
+starting with the factor, then use [`step_by`][iterator-step_by] with the same factor.
+Because the range has no upper bound, we then use [`take_while`][iterator-take_while] to
+stop the iteration when we reach the `limit`.
+
+To combine the multiples of all factors, we iterate the list of factors and use [`flat_map`][iterator-flat_map]
+on each factor's multiples. `flat_map` is a combination of [`map`][iterator-map] and [`flatten`][iterator-flatten];
+it maps each factor into its multiples, then flattens them all in a single sequence.
+
+To gather the multiples into a [`BTreeSet`][btreeset] which removes duplicates automatically,
+we use [`collect`][iterator-collect]. This powerful method can gather the values in an iterator
+into many different type of sequences; because of this, the compiler cannot infer the exact output
+type we're seeking. To help the compiler, we specify the desired output type using the so-called
+[turbofish] syntax: `::<>`. (Note that although we need to specify that the output needs to be
+a `BTreeSet`, we do not need to specify the set's _type_ - this can be inferred by the compiler.
+To let the compiler infer the set's type, we use the placeholder type specification: `_`.)
+
+Finally, calculating the sum of the remaining unique multiples in the set is easy: we can simply call [`sum`][iterator-sum].
+
+## Nightly Rust and `partition_dedup`
+
+Collecting the multiples in a [`BTreeSet`][btreeset] works, but might be expensive if we insert many multiples in it.
+If we're ready to use a Nightly Rust compiler, we can take advantage of a method that is not yet stabilized: [`partition_dedup`][slice-partition_dedup].
+This method removes consecutive duplicates from a slice. If the slice was sorted to begin with, then all duplicates are removed.
+Furthermore, it works in-place by moving duplicates to the end of the slice, thus not allocating new memory.
+
+If we use [`partition_dedup`][slice-partition_dedup], our code becomes:
+
+```rust
+#![feature(slice_partition_dedup)]
+
+pub fn sum_of_multiples(limit: u32, factors: &[u32]) -> u32 {
+    let mut multiples: Vec<_> = factors
+        .iter()
+        .filter(|&&factor| factor != 0)
+        .flat_map(|&factor| multiples(limit, factor))
+        .collect();
+
+    multiples.sort();
+    let (unique_multiples, _) = multiples.partition_dedup();
+    unique_multiples.iter().sum()
+}
+```
+
+Here, instead of using a [`BTreeSet`][btreeset], we simply collect the multiples into a [`Vec`][vec]. This is
+more efficient since `Vec` will allocate much less memory. Then, in order to have all duplicates removed
+by [`partition_dedup`][slice-partition_dedup], we first [`sort`][slice-sort] the multiples. (For an explanation
+of why we use [`sort`][slice-sort] and not [`sort_unstable`][slice-sort_unstable], see the documentation of [`sort`][slice-sort].)
```

Comment:
It's this part that explains it, I think (emphasis mine):

> Current implementation
>
> The current algorithm is an adaptive, iterative merge sort inspired by [timsort](https://en.wikipedia.org/wiki/Timsort). It is designed to be very fast in cases where the slice is nearly sorted, **or consists of two or more sorted sequences concatenated one after another**.

Here, because we're concatenating sorted sequences of multiples, it seems we're hitting a sweet spot in the algorithm. When I run benchmarks on my machine, `sort_unstable` is about 2x slower than `sort` for this use case.

Reply:
However, re-reading my text, the explanation doesn't seem enough. I think I have two options:

1. Remove the reference to `sort_unstable` entirely, since this discussion is a bit outside of the scope of the text
2. Add a bit more explanation (e.g., copy part of the documentation for `sort` in my text)

What do you think?

Reply:
I didn't know this! :smiley: So I think it would be a great addition. That little quote from the documentation with the explanation that our vector is already largely sorted makes total sense. I would also mention that you did actually do a benchmark (signalling to students that's the right thing to do). (No need to include the source of the benchmark, this is not a scientific paper.)